syserol/model.py

```diff
@@ -0,0 +1,144 @@
+""" Regression methods using Factorized Data. """
+import pandas as pd
+import numpy as np
+import tensorly as tl
+from sklearn.model_selection import KFold
+from sklearn.linear_model import ElasticNetCV, ElasticNet, LogisticRegressionCV
+from sklearn.model_selection import cross_val_predict
+from sklearn.metrics import r2_score, confusion_matrix
+from scipy.stats import zscore
+from tensorly.kruskal_tensor import kruskal_to_tensor
+from syserol.tensor import perform_CMTF
+from syserol.dataImport import createCube, importFunction, load_file, importGlycan
+
+
+def function_predictions(function="ADCD"):
+    """ Predict Function using Factorized Antigen Data"""
+    # import
+    cube, glyCube = createCube()
+    tensorFac, _, _ = perform_CMTF(cube, glyCube, 16)
+    func, _ = importFunction()
+
+    # arrange
+    df = pd.DataFrame(tensorFac[1][0])  # subjects x components matrix
+    df = df.join(func, how="inner")
+    df = df.dropna()
+    df_func = df[["ADCD", "ADCC", "ADNP", "CD107a", "IFNy", "MIP1b"]]
+    df_variables = df.drop(["subject", "ADCD", "ADCC", "ADNP", "CD107a", "IFNy", "MIP1b"], axis=1)
+
+    # Regression
+    X = df_variables
+    Y = df_func[function]
+    regr = ElasticNetCV(normalize=True, max_iter=10000)
+    model = regr.fit(X, Y)
+    Y_pred = cross_val_predict(ElasticNet(alpha=regr.alpha_, normalize=True, max_iter=10000), X, Y, cv=10)
+
+    print(model.coef_)
+    print(np.sqrt(r2_score(Y, Y_pred)))
+
+    return Y, Y_pred, np.sqrt(r2_score(Y, Y_pred))
+
+
+def subject_predictions():
+    """ Predict Subject Classifications using Factorized Antigen Data"""
+    cube, glyCube = createCube()
+    tensorFac, _, _ = perform_CMTF(cube, glyCube, 16)
+
+    # Assemble data
+    df = pd.DataFrame(tensorFac[1][0])  # subjects x components matrix
+    subj = load_file("meta-subjects")
+    df = df.join(subj, how="inner")
+    df = df.dropna()
+
+    # Subset, Z score
+    df_class = df[["class.cp", "class.nv"]]
+    df_variables = df.drop(["subject", "class.etuv", "class.cp", "class.nv"], axis=1)
+    df_variables = df_variables.apply(zscore)
+
+    # Predict Controller vs. Progressor
+    Y1 = df_class["class.cp"]
+    Y1 = (Y1 == "controller").astype(int)  # controllers are 1s, progressors are 0s
+    X1 = df_variables
+
+    Y_pred1 = cross_val_predict(LogisticRegressionCV(max_iter=10000), X1, Y1)
```

Comment:
Also, this code seems older—can we merge unrelated parts on a separate PR. It's really, really, really helpful for PRs to be about exactly one change.

Reply:
Yes, this code is the original code I put in this PR - trying to run our own predictions after performing factorization. This PR has been open for a while now with lots of things being added to it because it's all been related to the syserol.model file which has so far only existed on this branch. It's had a lot of building because we've used a lot of different strategies for analyzing how well our model predicts and how we predict in the first place. I can put this older code on a different PR though